File: from-0001-to-0100/0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        ans = 0
        size = len(height)
        left_max = [0] * size
        right_max = [0] * size
        
        left_max[0] = height[0]
        for i in range(1, size):
            left_max[i] = max(left_max[i-1], height[i])
        
        right_max[-1] = height[-1]
        for i in range(size-2, -1, -1):
            right_max[i] = max(right_max[i+1], height[i])
            
        # print(left_max)
        # print(right_max)
            
        for i in range(1, size-1):
            ans += (min(left_max[i], right_max[i]) - height[i])
            
        return ans

File: from-0001-to-0100/0042-trapping-rain-water/solution.go
func trap(height []int) int {
    n := len(height)
    if n == 0 {
        return 0
    }
    result := 0
    leftMax, rightMax := make([]int, n), make([]int, n)
    
    leftMax[0] = height[0]
    for i := 1; i < n; i ++ {
        leftMax[i] = max(height[i], leftMax[i - 1])
    }
    rightMax[n - 1] = height[n - 1]
    for i := n - 2; i >= 0; i -- {
        rightMax[i] = max(height[i], rightMax[i + 1])
    }

    for i := 1; i < n - 1; i ++ {
        result += min(leftMax[i], rightMax[i]) - height[i]
    }
    return result
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}